### PR TITLE
[CodeExtractor] Refactor extractCodeRegion, fix alloca emission.

### PR DESCRIPTION
**Pull request by Meinersbur**
Reorganize the code into phases:

 * Analyze/normalize
 * Create extracted function prototype
 * Generate the new function's implementation
 * Generate call to new function
 * Connect call to original function's CFG

The motivation is #114669 to optionally clone the selected code region into the new function instead of moving it. The current structure made it difficult to add such functionality since there was no obvious place to do so, not made easier by some functions doing more than their name suggests. For instance, constructFunction modifies code outside the constructed function, but also function properties such as setPersonalityFn are derived somewhere else. Another example is emitCallAndSwitchStatement, which despite its name also inserts stores for output parameters.

Many operations also implicitly depend on the order they are applied which this patch tries to reduce. For instance, ExtractedFuncRetVals becomes the list exit blocks which also defines the return value when leaving via that block. It is computed early such that the new function's return instructions and the switch can be generated independently. Also, ExtractedFuncRetVals is combining the lists ExitBlocks and OldTargets which were not always kept consistent with each other or NumExitBlocks. The method recomputeExitBlocks() will update it when necessary.

The coding style partially contradict the current coding standard. For instance some local variable start with lower case letters. I updated some, but not all occurrences to make the diff match at least some lines as unchanged.

The patch [D96854](https://reviews.llvm.org/D96854) introduced some confusion of function argument indexes this is fixed here as well, hence the patch is not NFC anymore. Tested in modified CodeExtractorTest.cpp. Patch [D121061](https://reviews.llvm.org/D121061) introduced AllocationBlock, but not all allocas were inserted there.

Efectively includes the following fixes:
1. https://github.com/llvm/llvm-project/commit/ce73b1672a6053d5974dc2342881aac02efe2dbb
2. https://github.com/llvm/llvm-project/commit/4aaa92578686176243a294eeb2ca5697a99edcaa
3. Missing allocas, still unfixed

Originally submitted as https://reviews.llvm.org/D115218

### CODE REVIEW

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp
Comment (tblah) [inline llvm/lib/Transforms/Utils/CodeExtractor.cpp:1896]: I don't really understand why this behavior is valid/safe, but I can see you are just replicating the old behavior so you're right not to change it in this patch.

**Review by tblah — COMMENTED** [review]
I have a few questions to clarify bits I don't understand. I'm relatively new to the middle end so those aren't necessarily things that need to be fixed in the patch.

Overall this is a big cleanup with lots of small improvements in the modified code. The parts I understand it look good to me but I wouldn't want to approve it myself without somebody with more expertise taking a look too. Ideally this would be split into a lot of smaller patches to make it easier to review, but I understand how that can quickly become difficult for such a big refactoring.

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp
Comment (Meinersbur) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1430]: So `EntryFreq` and `ExitWeights` are computed together.

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp
Comment (Meinersbur) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1637]: Because the original code does. I don't know why. Undo the inversion and see regression tests fail.

Admittedly, the comment on `SwitchCases` is a simplification. I will update it.

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp
Comment (Meinersbur) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1904]: "or something" indicates the [original implementor](https://github.com/llvm/llvm-project/commit/7f1c7ede5b134845a0f98d38dcfe253ebd9991b4) doesn't know either (the exception model might not have been fixed by then), and there is no test triggering this. There were some [recent changes](https://github.com/llvm/llvm-project/commit/ff63d628c46459437df21e09f70993685eba4bc3) in that added the unreachable. I think if there is no exit block, there is no possibility to actually return, and the case should be simplifyable to this:
```suggestion
assert(newFunction->doesNotReturn());
(void)new UnreachableInst(Context, TheSwitch->getIterator());
```
It passes all regression tests, but the subject of this PR is the refactoring.

**Comment by Meinersbur** [comment]
> Overall this is a big cleanup with lots of small improvements in the modified code. The parts I understand it look good to me but I wouldn't want to approve it myself without somebody with more expertise taking a look too. 

Such person does not exist. If you are not going to approve these changes, we can save both of us some time and abandon this PR again right here and now and declare CodeExtractor abandoned code without maintainer. I already tried 3 years ago. IMHO CodeExtractor in the current form is a bug breeding ground. 3, possibly 4, bugs fixed by just this refactoring.

**Comment by kiranchandramohan** [comment]
Thanks @Meinersbur for this refactoring. Would it be OK to upstream this incrementally in small patches for ease of review? I have taken one small change from this patch and created https://github.com/llvm/llvm-project/pull/115078. 
Note: This is only to demonstrate and express my willingness to help.

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp
Comment (tblah) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1904]: Agreed. Lets keep this PR to just the refactoring

**Review by tblah — APPROVED** [review]
Thanks for answering my questions. In the absence of anyone else to review this patch, it looks good to me.

Thanks for taking the time to clean up this code and fix the longstanding bugs.

File: llvm/include/llvm/Transforms/Utils/CodeExtractor.h
Comment (goldsteinn) [inline llvm/include/llvm/Transforms/Utils/CodeExtractor.h:123]: nit: I think this should be "ExitBlocks"

**Comment by kiranchandramohan** [comment]
> Thanks @Meinersbur for this refactoring. Would it be OK to upstream this incrementally in small patches for ease of review? I have taken one small change from this patch and created #115078. Note: This is only to demonstrate and express my willingness to help.

In a similiar spirit as  #115078 we could trivially refactor applyFirstDebugLoc in an NFC patch.  Next we could just correct `emitCallAndSwitchStatement` to do something like `emitReplacerCall`. However, I don't know whether this will be a worthwhile exercise. Given that you have an approval, would you prefer to go ahead and submit it? Or would you like me to help here with making smaller incremental refactoring?

File: llvm/include/llvm/Transforms/Utils/CodeExtractor.h
Comment (Meinersbur) [reply llvm/include/llvm/Transforms/Utils/CodeExtractor.h:123]: I intentionally named it `SwitchCases` to indicate that it is not just an unordered list of exist blocks, but the order is used to determine the extracted function's return values and the replacement switch statement. I am not married to the name (could also be `ExtractedFunctionReturnValues`), but I honestly think that `ExitBlocks` would be a misleading name.

Despite this, do you still prefer `ExitBlocks`?

Comment (goldsteinn) [reply llvm/include/llvm/Transforms/Utils/CodeExtractor.h:123]: Just any name that doesn't imply it's only used by a switch instruction.

**Comment by Meinersbur** [comment]
@kiranchandramohan I will do refactorings like `getSwitchType` and `applyFirstDebugLoc` separately if requested, put consider that patch series significantly increases the churn of splitting, creating, updating, formatting, merging, and testing each individual PR.

In contrast I have no idea how `emitReplacerCall` could be isolated. The problem with the current state of CodeExtractor is that all the function calls are intermingled such that that `emitCallAndSwitchStatement` does more and at the same time less than `emitReplacerCall` does. Those things would need to moved to somewhere else and at the same time code that does things that `emitReplacerCall` does but not `emitCallAndSwitchStatement` deleted elsewhere. That is, the order in which things are done are neither consistent with the old order, nor with the refactored order, needing additional intermitted fixes (or intentional preservation of bugs) that would be eventually removed again, increasing the amount of work for reviewers. I am not sure I would even be able to get such an intermediate state working correctly.

Consider that one can review this code like new code by ignoring the previous state. If we are not able to do so we would never be able to add any new feature. Imagine I would add `CodeExtractor2.cpp` and in a follow-up redirect uses of `CodeExtractor` to `CodeExtractor2`.

**Comment by kiranchandramohan** [comment]
> @kiranchandramohan I will do refactorings like `getSwitchType` and `applyFirstDebugLoc` separately if requested, put consider that patch series significantly increases the churn of splitting, creating, updating, formatting, merging, and testing each individual PR.
> 
> In contrast I have no idea how `emitReplacerCall` could be isolated. The problem with the current state of CodeExtractor is that all the function calls are intermingled such that that `emitCallAndSwitchStatement` does more and at the same time less than `emitReplacerCall` does. Those things would need to moved to somewhere else and at the same time code that does things that `emitReplacerCall` does but not `emitCallAndSwitchStatement` deleted elsewhere. That is, the order in which things are done are neither consistent with the old order, nor with the refactored order, needing additional intermitted fixes (or intentional preservation of bugs) that would be eventually removed again, increasing the amount of work for reviewers. I am not sure I would even be able to get such an intermediate state working correctly.
> 
> Consider that one can review this code like new code by ignoring the previous state. If we are not able to do so we would never be able to add any new feature. Imagine I would add `CodeExtractor2.cpp` and in a follow-up redirect uses of `CodeExtractor` to `CodeExtractor2`.

I was thinking of the smaller refactoring approach since I thought unfamiliar reviewers would struggle with making sense of the huge change. But since @tblah has gone ahead and made sense of it all, this does not hold.

I agree that going ahead with this patch (as it is) is the best way forward and the pragmatic approach.

File: llvm/lib/Transforms/Utils/CodeExtractor.cpp
Comment (aeubanks) [inline llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474]: after this commit, we're seeing msan failures saying there's a use of uninitialized memory on this line

```
    #0 0x7f1a2b16a7a1 in llvm::CodeExtractor::extractCodeRegion(llvm::CodeExtractorAnalysisCache const&, llvm::SetVector<llvm::Value*, llvm::SmallVector<llvm::Value*, 0u>, llvm::DenseSet<llvm::Value*, llvm::DenseMapInfo<llvm::Value*, void>>, 0u>&, llvm::SetVector<llvm::Value*, llvm::SmallVector<llvm::Value*, 0u>, llvm::DenseSet<llvm::Value*, llvm::DenseMapInfo<llvm::Value*, void>>, 0u>&) llvm-project/llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474:3
    #1 0x7f1a2b168abf in llvm::CodeExtractor::extractCodeRegion(llvm::CodeExtractorAnalysisCache const&) llvm-project/llvm/lib/Transforms/Utils/CodeExtractor.cpp:1377:10
    #2 0x7f1a2f777242 in llvm::OpenMPIRBuilder::finalize(llvm::Function*) llvm-project/llvm/lib/Frontend/OpenMP/OMPIRBuilder.cpp:719:38
    #3 0x7f1a3522aa86 in (anonymous namespace)::OpenMPIRBuilderTest_ParallelNested_Test::TestBody() llvm-project/llvm/unittests/Frontend/OpenMPIRBuilderTest.cpp:921:14
```

Comment (aeubanks) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474]: @Meinersbur

Comment (Meinersbur) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474]: I suspect it to be `StructTy` which is not initialized if not using a struct. Would not be used in that case either. Working on a fix...

Comment (Meinersbur) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474]: This was fixed already in 9b6b9d39030f92ea20d8f38ef37305e99cfc6d60 which was committed before your comment.

It would be good if you had linked the buildbot.

Comment (aeubanks) [reply llvm/lib/Transforms/Utils/CodeExtractor.cpp:1474]: sorry, this was in some internal non-public bots. yes that commit fixes this, thanks for the pointer

**Pull request merged**
